privacyidea/lib/tokens/tiqrtoken.py

```diff
@@ -384,6 +391,10 @@ def create_challenge(self, transactionid=None, options=None):
         # Depending on the OCRA-SUITE we create the challenge
         os = OCRASuite(ocrasuite)
         challenge = os.create_challenge()
+        if int(challenge) >= 2147483648:
+            # TIQR app has problems with value greater the 2^31
+            challenge = '1' + challenge[1:]
```

Comment:
Should we downshift all values >=2147483648? Then there are certain values with a higher probability (1147483648, 1147483649 have double the probability like 1147483647, 1147483646) or should we rather reshuffle:
~~~~python
challenge = 2147483648
while challenge >= 2147483648:
    challenge = os.create_challenge()
~~~~

Reply:
My concern was that creating a challenge is an "expensive" operation and opted for this approach. We can also use a flip coin approach.  Replace the first digit with a `0` or a `1`.   Then the probability will lower a bit.

@cornelinux you make the decision what is the right approach and I will implement it.

Reply:
I looked into it again. The default OCRAsuite for the TiQR token  is currently:
"OCRA-1:HOTP-SHA1-6:QN10"

This is a ten digit number ranging from 0000000001 to 999999999.
So roughly 78% are bigger than 2147483648.
I am not quite sure of the limitations here. Is the the TiQR protocol in itself or the implementation of the App? Maybe it might make more sense to simply change the default_suite, which defines the challenge to OCRA-1:HOTP-SHA1-6:QN9 (only 9 digits)

Reply:
@basvandervlies where do you get the 31bit limitation from?

Reply:
Hm, maybe we should add TiQR to the privacyIDEA authenticator to get rid of this limitation.
I think there is no need to change this in the code. You could also specify another OCRASuite (like QN9) in the WebUI in Config->Token->TiQR

Reply:
@cornelinux I just stumbled on this by Trial and Error because the default is `OCRA-1:HOTP-SHA1-6:QN10` for PrivacyIdea.  See also:
 * https://community.privacyidea.org/t/tiqr-response-problems/1100/8

I changed the `QN` part of the ocrasuite to `9`, `09`, `8` and `08` and they do not work at all. There is something wrong the official app.   I can only test the IOS app. I have updated the issue:
 * https://github.com/SURFnet/tiqr-client-ios/issues/39

The only code that works for me is `QN:10` with the check of value is not too big

Reply:
I am also in contact with the TIQR app developers and our sister organization SURFnet. They say the app is tested with suite: `OCRA-1:HOTP-SHA1-6:QH10-S512`

But the `S` is not implemented in privacyidea:
```
NotImplementedError: OCRA Session not implemented, yet.
```

Also tested other suites without the `S` option, but they also fail

Reply:
We have now tested with `OCRA-1:HOTP-SHA1-6:QH10`, which seems to work. We had one enrolled token that failed 100% of the time, but after that we couldn't reproduce it with other tokens. A 10 digit hexadecimal number has a high probability to be larger than 31 bit, so the same limitation does not seem to apply.

Reply:
@martijnkruiten, @basvandervlies What now?
I would rather not add this challenge-reduce-code into the upstream code, since it seems to be a special fix for the TiQR app. So I think there are two possibilities:

1. Identify the an OCRASuite that works for you and configure it that way, or
2. run the "challenge-reduce-code"  only if a special config value in pi.cfg is set like
    ~~~~python
    if int(challenge) >= 2147483648 and get_app_config_value("PI_REDUCE_TIQR_CHALLENGE") is True:
            # TIQR app has problems with value greater the 2^31
            challenge = '1' + challenge[1:]
    ~~~~
3. Have the TiQR developer fix the app

Reply:
@cornelinux I have removed the fix for the this special case. Its indeed depended on the OCRAsuite used. It can be a bug in the app or in the server code with conversion.  The developpers of the APP will also take a look at it.

We have now switched to: `OCRA-1:HOTP-SHA1-6:QH10` and that works much better. But we encountered something else that the length of the challenge is incorrect. Is must be 10 instead of 20. That is the other improvement. But my colleague just informed my and we will put it in the TIQR token module

Reply:
Yes, see my other comment. This must be fixed in lib/tokens/ocra.py.